.readthedocs.yml

```diff
@@ -4,6 +4,9 @@ build:
   os: "ubuntu-22.04"
   tools:
     python: "mambaforge-22.9"
+  jobs:
+    pre_install:
+      - bash ci/build_docs_pre_install.sh
```

Comment:
There is no `pip`-installable library `libucx`... it needs to be either `libucx-cu11` or `libucx-cu12`.

Readthedocs builds are installing directly from the `pyproject.toml` checked into source control, which
doesn't have those suffixes added.

https://github.com/rapidsai/ucx-py/blob/03c864b790a276e6732e00b962c6cb4d143eb786/.readthedocs.yml#L8-L11

Resulting in this:

> INFO: pip is looking at multiple versions of ucx-py to determine which version is compatible with other requirements. This could take a while.
> ERROR: Could not find a version that satisfies the requirement libucx<1.16,>=1.15.0 (from ucx-py) (from versions: none)
> ERROR: No matching distribution found for libucx<1.16,>=1.15.0

([example build link](https://readthedocs.org/projects/ucx-py/builds/24296596/))

`readthedocs` doesn't allow customizing the `pip install` with arbitrary flags, e.g. by adding `--no-deps` ([docs](https://docs.readthedocs.io/en/stable/config-file/v2.html#python)).

So I think this pre-install script to fix the suffixes is the best option to keep those builds working. For more details on how it works, see https://docs.readthedocs.io/en/stable/build-customization.html.

Reply:
I agree, that seems a bit cleaner to me.

Did you try `PIP_NO_DEPENDENCIES`? I don't remember which is the right name (or maybe both work), and also the behavior of true/false for pip's `*_NO_*` variables can be confusing so it's worth double-checking all the options with the setting that you proposed in the env.yaml before we set this on RTD. If all else fails, though, I'm fine with setting on RTD.

Reply:
> Did you try `PIP_NO_DEPENDENCIES`?

I didn't, but that'd just take a minute. I'll do that.

The `pip` docs mention both of them: https://pip.pypa.io/en/latest/cli/pip_install/#cmdoption-no-deps

Reply:
OK cool. If that doesn't work, I'm fine with setting the flags appropriately in our RTD env.

Reply:
Ok yeah, none of these variations worked:

```yaml
variables:
  - PIP_NO_DEPENDENCIES: true
  
variables:
  - PIP_NO_DEPENDENCIES: "true"

variables:
  - PIP_NO_DEPENDENCIES: "on"

variables:
  - PIP_NO_DEPENDENCIES: "off"

variables:
  - PIP_NO_DEPENDENCIES: 1
```

@pentschev could you and set environment variable `PIP_NO_DEPENDENCIES` to "True" for readthedocs builds here, following https://docs.readthedocs.io/en/stable/guides/environment-variables.html? One you do that, I'll push a commit to test.

Reply:
Done.

Reply:
@jameslamb I invited you to be a RTD maintainer as well, this might make things easier for you to test.

Reply:
Excellent, thank you!! Testing right now.

Reply:
Alright that failed ([build link](https://readthedocs.org/projects/ucx-py/builds/24324188/)), I think because the environment variable you set up was marked "private".

From https://docs.readthedocs.io/en/stable/environment-variables.html#custom-env-var-exceptions

> *Custom environment variables that are not marked as Public will not be available in [pull request builds](https://docs.readthedocs.io/en/stable/pull-requests.html)*

I just created a new public one and kicked off a rebuild... unfortunately that failed too 😭 ([build link](https://readthedocs.org/projects/ucx-py/builds/24324249/))

Thanks for adding me as a maintainer, I'll use those permissions to keep trying stuff.

Reply:
Ahhhhh I forgot that `PIP_NO_DEPENDENCIES` does not stop `pip install` from trying to download **build** dependencies.

```text
Processing /home/docs/checkouts/readthedocs.org/user_builds/ucx-py/checkouts/1041
  Installing build dependencies: started
  Installing build dependencies: finished with status 'error'
  error: subprocess-exited-with-error
  
  × pip subprocess to install build dependencies did not run successfully.
  │ exit code: 1
  ╰─> [4 lines of output]
      Collecting cython>=3.0.0
        Downloading Cython-3.0.10-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (3.2 kB)
      ERROR: Could not find a version that satisfies the requirement libucx==1.15.0 (from versions: none)
      ERROR: No matching distribution found for libucx==1.15.0
      [end of output]
```

([build link](https://readthedocs.org/projects/ucx-py/builds/24324302/))

I just added `PIP_NO_BUILD_ISOLATION="True"` to the environment as well... still didn't work 😞 

([build link](https://readthedocs.org/projects/ucx-py/builds/24324476/))

I have one other idea, will post in a second.

Reply:
To summarize.... these things did not work:

* set env variable `PIP_NO_DEPS="True"` in readthedocs builds
* set env variable `PIP_NO_BUILD_ISOLATION="True"` in readthedocs builds
* set both of the above in readthedocs builds
* set any combination of those environment variables in `variables: ` block in `builddocs.yml` conda env
* pass `pip: - --no-deps ../../` via `builddocs.yml` conda env
* use a real, suffixed dependency name like `libucx-cu11>=1.15.0` in `pyproject.toml` + add `--extra-index-url=https://pypi.anaconda.org/rapidsai-wheels-nightly/simple` in `builddocs.yml` conda env

That last one failing revealed a more significant issue, not limited to just docs building...

**calling `libucx.load_library()` unconditionally means `ucx-py` will not be importable unless the process has access to a GPU** 😱 

```text
RuntimeError: The CUDA driver library libcuda.so.1 was not found on your system. This library cannot be provided by the libucx wheel and must be installed separately.
```

([build link](https://readthedocs.org/projects/ucx-py/builds/24325258/))

The docs DO BUILD successfully on the current state of this branch 🎉 ([build link](https://readthedocs.org/projects/ucx-py/builds/24325407/)), but only because I added behavior to work around not having a GPU available.

So @pentschev @vyasr that's a design decision for you.

When someone tries to `import ucp` without access to a GPU, what should happen?

* the import should fail
* the import should warn but succeed
* something else

Reply:
Summarizing our offline conversations:

* yes `ucx-py` is expected to be usable without a GPU
* the `libucx` wheels should stop raising exceptions when `libucx.load_library()` is called in an environment without a GPU

In response, we published a new `libucx==1.15.0.post1` removing those exceptions: https://github.com/rapidsai/ucx-wheels/pull/5

Tested it with `ucxx` on systems with and without a GPU, and it seemed to work well: https://github.com/rapidsai/ucx-wheels/pull/5#issuecomment-2103619706

Those new wheels helped the docs builds here get further, but those builds exposed another issue... as of the changes in this PR, `ucx-py` was compiling with the **system-installed headers** (e.g. `/usr/include/ucm`) but the **wheel-provided shared libraries** (e.g. `site-packages/libucx/lib/libucm.so.0`).

That showed up in "symbol not found" issues like this:

```text
OSError: ${HOME}/conda/1041/lib/python3.12/site-packages/libucx/lib/libucs_signal.so: undefined symbol: ucs_debug_is_error_signal
```

([build link](https://readthedocs.org/projects/ucx-py/builds/24342353/))

I just pushed a change that fixes that: https://github.com/rapidsai/ucx-py/pull/1041/commits/eba110fe2116978e9488b46b7d884ab6b5ffb876

And looks like the docs builds are happy and docs are rendering correctly ([RTD build link](https://readthedocs.org/projects/ucx-py/builds/24343567/)).

Reply:
Good catch on the headers! I forgot to check on those. Glad you found this issue.

Reply:
Update for those finding this thread from search or other links.... thanks to @bdice 's suggestion in https://github.com/rapidsai/ucx-py/pull/1044#discussion_r1636690394, we were able to get this working without messing with environment variables or the conda environment YAML file, like this in `.readthedocs.yaml`:

```yaml
    post_create_environment:
      - |
        pip install \
          --extra-index-url "https://pypi.anaconda.org/rapidsai-wheels-nightly/simple/" \
          -C rapidsai.disable-cuda=true \
          -C rapidsai.matrix-entry="cuda=12.2" \
          .
```